internal/utilities/hibpcache.go

```diff
@@ -0,0 +1,69 @@
+package utilities
+
+import (
+	"context"
+	"sync"
+
+	"github.com/bits-and-blooms/bloom/v3"
+)
+
+type HIBPBloomCache struct {
+	sync.RWMutex
+
+	n      uint
+	items  uint
+	filter *bloom.BloomFilter
+}
+
+func NewHIBPBloomCache(n uint, fp float64) *HIBPBloomCache {
+	cache := &HIBPBloomCache{
+		n:      n,
+		filter: bloom.NewWithEstimates(n, fp),
+	}
+
+	return cache
+}
+
+func (c *HIBPBloomCache) Cap() uint {
+	return c.filter.Cap()
+}
+
+func (c *HIBPBloomCache) Add(ctx context.Context, prefix []byte, suffixes [][]byte) error {
+	c.Lock()
+	defer c.Unlock()
+
+	c.items += uint(len(suffixes))
+
+	if c.items > (4*c.n)/5 {
+		// clear the filter if 80% full to keep the actual false
+		// positive rate low
+		c.filter.ClearAll()
+
+		// reduce memory footprint when this happens
+		c.filter.BitSet().Compact()
+
+		c.items = uint(len(suffixes))
+	}
+
+	var combined [40]byte
```

Comment:
Maybe missed this (and it might be from the calculator) but could we leave a comment about where the number 40 is derived from